node/network/bridge/src/lib.rs

```diff
@@ -0,0 +1,517 @@
+				Some(Action::PeerDisconnected(remote))
+			} else {
+				None
+			}
+		}
+		Some(NetworkEvent::NotificationsReceived { remote, messages }) => {
+			let v: Result<Vec<_>, _> = messages.iter()
+				.filter(|(engine_id, _)| engine_id == &POLKADOT_ENGINE_ID)
+				.map(|(_, msg_bytes)| WireMessage::decode(&mut msg_bytes.as_ref()))
+				.collect();
+
+			match v {
+				Err(_) => Some(Action::ReportPeer(remote, MALFORMED_MESSAGE_COST)),
+				Ok(v) => if v.is_empty() {
+					None
+				} else {
+					Some(Action::PeerMessages(remote, v))
+				}
+			}
+		}
+	}
+}
+
+fn construct_view(live_heads: &[Hash]) -> View {
+	View(live_heads.iter().rev().take(MAX_VIEW_HEADS).cloned().collect())
+}
+
+async fn dispatch_update_to_all(
+	update: NetworkBridgeEvent,
+	event_producers: impl IntoIterator<Item=&fn(NetworkBridgeEvent) -> AllMessages>,
+	ctx: &mut impl SubsystemContext<Message=NetworkBridgeMessage>,
+) -> polkadot_subsystem::SubsystemResult<()> {
+	// collect messages here to avoid the borrow lasting across await boundary.
+	let messages: Vec<_> = event_producers.into_iter()
+		.map(|producer| producer(update.clone()))
+		.collect();
+
+	ctx.send_messages(messages).await
+}
+
+async fn update_view(
+	peers: &HashMap<PeerId, PeerData>,
+	live_heads: &[Hash],
+	net: &impl Network,
+	local_view: &mut View,
+) -> Option<NetworkBridgeEvent> {
+	let new_view = construct_view(live_heads);
+	if *local_view == new_view { return None }
+	*local_view = new_view.clone();
+
+	let message = WireMessage::ViewUpdate(new_view.clone()).encode();
+
+	let write_all = peers.keys().cloned().map(|peer| {
+		net.write_notification(peer, message.clone())
+	});
+
+	future::join_all(write_all).await;
+
+	Some(NetworkBridgeEvent::OurViewChange(local_view.clone()))
+}
+
+async fn run_network(net: impl Network, mut ctx: impl SubsystemContext<Message=NetworkBridgeMessage>) {
+	let mut event_stream = net.event_stream().fuse();
+
+	// Most recent heads are at the back.
+	let mut live_heads = Vec::with_capacity(MAX_VIEW_HEADS);
+	let mut local_view = View(Vec::new());
+
+	let mut peers = HashMap::new();
+	let mut event_producers = HashMap::new();
+
+	loop {
+		let action = {
+			let subsystem_next = ctx.recv().fuse();
+			let mut net_event_next = event_stream.next().fuse();
+			futures::pin_mut!(subsystem_next);
+
+			let action = futures::select! {
+				subsystem_msg = subsystem_next => Some(action_from_overseer_message(subsystem_msg)),
+				net_event = net_event_next => action_from_network_message(net_event),
+			};
+
+			match action {
+				Some(a) => a,
+				None => continue,
+			}
+		};
+
+		match action {
+			Action::RegisterEventProducer(protocol_id, event_producer) => {
+				// insert only if none present.
+				event_producers.entry(protocol_id).or_insert(event_producer);
+			}
+			Action::SendMessage(peers, protocol, message) => {
+				let message = WireMessage::ProtocolMessage(protocol, message).encode();
+
+				for peer in peers.iter().skip(1).cloned() {
```

Comment:
It prevents a `clone` if we are sending only to one peer

Reply:
I was curious whether this *loop unrolling* would be more performant than the optimizations rustc and llvm do already.

I wrote a small benchmark https://github.com/mxinden/clone-trick/blob/master/benches/bench.rs distributing a message of 1 KB, 10 KB, 100 KB and 1 MB among 20 peers.

![image](https://user-images.githubusercontent.com/7047859/85850746-af825180-b7ad-11ea-92e5-647eef54d578.png)

The manual loop unrolling does have a performance impact, but a small one. While the simple implementation takes 3.9 us to distribute a 10 KB message to 20 peers  the optimized implementation takes 3.6 us (on a `i7-8550U`).

I would expect that if `write_notification` accepts some type that allows sharing of the underlying allocation (e.g. `Bytes` or `Arc<Vec<u8>>` like we do for [`NotificationsReceived`](https://github.com/paritytech/substrate/blob/master/client/network/src/protocol/event.rs#L75) the entire memory allocation overhead would vanish.

@rphmeier in case you want to keep the optimization in here (e.g. for the impact it has in case there is only a single peer and thus no `clone` needed), would you mind adding a comment describing why this is not a single for loop?

Reply:
It's a memory optimization, not a performance one. Computers are pretty fast, but if I'm sending a 10MiB message to 20 peers I'd rather keep memory usage down by 5%.

Reply:
It would also be addressed by the lower level APIs accepting a set of peers to send a message to.

Reply:
> but if I'm sending a 10MiB message to 20 peers I'd rather keep memory usage down by 5% while this is buffered

Are you suggesting that this optimization keeps the [resident set size](https://en.wikipedia.org/wiki/Resident_set_size) in memory low? How does it do that?

Say we send a message to 20 peers. With the un-optimized implementation the message would be cloned 20 times, thus for a short amount of time the message would exist 21 times in memory. But this 21st (useless) copy of the message is dropped right at the end of the function scope, thus this optimization only reduces the total resident set size for the duration of the function scope, which I would guess is negligible.

Reply:
Well, every little bit counts